<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Dataset preparation
This step performs all preparation necessary to perform feamiR analysis, taking a set of mRNAs, a set of miRNAs and an interaction dataset and creating corresponding positive and negative datasets for ML modelling. — preparedataset • feamiR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Dataset preparation
This step performs all preparation necessary to perform feamiR analysis, taking a set of mRNAs, a set of miRNAs and an interaction dataset and creating corresponding positive and negative datasets for ML modelling. — preparedataset" />
<meta property="og:description" content="PLEASE NOTE:
This analysis is run in Python so python must be installed and location specified if not on PATH.
Both sreformat and PaTMaN must also be installed and path specified if not on PATH.
Python &amp;gt;= 3.6 is required to use the neccesary packages.
The Python component required the following libraries: os, Bio, gtfparse, pandas, numpy, math, scipy.stats, matplotlib.pyplot, seaborn as sns, statistics, logging. Please ensure these are installed for the verison of Python you supply." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">feamiR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Dataset preparation
This step performs all preparation necessary to perform feamiR analysis, taking a set of mRNAs, a set of miRNAs and an interaction dataset and creating corresponding positive and negative datasets for ML modelling.</h1>
    
    <div class="hidden name"><code>preparedataset.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>PLEASE NOTE:
This analysis is run in Python so python must be installed and location specified if not on PATH.
Both sreformat and PaTMaN must also be installed and path specified if not on PATH.
Python &gt;= 3.6 is required to use the neccesary packages.
The Python component required the following libraries: os, Bio, gtfparse, pandas, numpy, math, scipy.stats, matplotlib.pyplot, seaborn as sns, statistics, logging. Please ensure these are installed for the verison of Python you supply.</p>
    </div>

    <pre class="usage"><span class='fu'>preparedataset</span><span class='op'>(</span>
  pythonversion <span class='op'>=</span> <span class='st'>"python"</span>,
  mRNA_3pUTR <span class='op'>=</span> <span class='st'>""</span>,
  miRNA_full <span class='op'>=</span> <span class='st'>""</span>,
  interactions <span class='op'>=</span> <span class='st'>""</span>,
  annotations <span class='op'>=</span> <span class='st'>""</span>,
  fullchromosomes <span class='op'>=</span> <span class='st'>""</span>,
  seed <span class='op'>=</span> <span class='fl'>1</span>,
  nonseed_miRNA <span class='op'>=</span> <span class='fl'>0</span>,
  flankingmRNA <span class='op'>=</span> <span class='fl'>0</span>,
  UTR_output <span class='op'>=</span> <span class='st'>""</span>,
  chr <span class='op'>=</span> <span class='st'>""</span>,
  o <span class='op'>=</span> <span class='st'>"feamiR_"</span>,
  positiveset <span class='op'>=</span> <span class='st'>""</span>,
  negativeset <span class='op'>=</span> <span class='st'>""</span>,
  sreformatpath <span class='op'>=</span> <span class='st'>"sreformat"</span>,
  patmanpath <span class='op'>=</span> <span class='st'>"patman"</span>,
  patmanoutput <span class='op'>=</span> <span class='st'>""</span>,
  minvalidationentries <span class='op'>=</span> <span class='fl'>40</span>,
  num_runs <span class='op'>=</span> <span class='fl'>100</span>,
  check_python <span class='op'>=</span> <span class='cn'>T</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pythonversion</th>
      <td><p>File path for installed Python version (default: python)</p></td>
    </tr>
    <tr>
      <th>mRNA_3pUTR</th>
      <td><p>Fasta file of only 3'UTRs, with gene name as name attribute (e.g. Serpinb8)</p></td>
    </tr>
    <tr>
      <th>miRNA_full</th>
      <td><p>Fasta file of full mature miRNA hairpins, with miRNA ID as name attribute (e.g. hsa-miR-576-3p)</p></td>
    </tr>
    <tr>
      <th>interactions</th>
      <td><p>CSV file containing only validated interactions between miRNA and mRNA (e.g. from miRTarBase). Must have columns miRNA (e.g. hsa-miR-576-3p), Target Gene (e.g. Serpinb8) and optionally Experiments (e.g. qRT-PCR) and/or Support Type (with values Functional MTI, Functional MTI (Weak), Non-Functional MTI, Non-Functional MTI (Weak))</p></td>
    </tr>
    <tr>
      <th>annotations</th>
      <td><p>GTF file (e.g. from Ensembl) with attributes seqname (chromosome), feature (with 3'UTRs labelled exactly 'three_prime_utr'), transcript_id, gene_id and gene_name matching fullchromosomes and interactions</p></td>
    </tr>
    <tr>
      <th>fullchromosomes</th>
      <td><p>Fasta file (e.g. top level file from Ensembl) containing full sequence for each chromosome with name as chromosome (e.g. 1, matching seqname from annotations)</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Binary, 1 if full miRNA seed features should be included in statistical analysis. Default: 1.</p></td>
    </tr>
    <tr>
      <th>nonseed_miRNA</th>
      <td><p>Binary, 1 if full miRNA features should be included in statistical analysis. Seed features are always included. Default: 0.</p></td>
    </tr>
    <tr>
      <th>flankingmRNA</th>
      <td><p>Binary, 1 if flanking region mRNA features should be included in statistical analysis. Seed features are always included. Default: 0.</p></td>
    </tr>
    <tr>
      <th>UTR_output</th>
      <td><p>String. File name 3'UTR fasta file should be saved as (when annotations and full chromosomes files are supplied)</p></td>
    </tr>
    <tr>
      <th>chr</th>
      <td><p>Number of chromosomes for species in question.</p></td>
    </tr>
    <tr>
      <th>o</th>
      <td><p>Output prefix for any files created and saved.</p></td>
    </tr>
    <tr>
      <th>positiveset</th>
      <td><p>CSV file containing validated pairs of miRNAs and mRNAs as output by initial stage of analysis. If positiveset and negative set are input, analysis begins at final statistical analysis stage.</p></td>
    </tr>
    <tr>
      <th>negativeset</th>
      <td><p>CSV file containing non-validated pairs of miRNAs and mRNAs as output by initial stage of analysis. If positiveset and negative set are input, analysis begins at final statistical analysis stage.</p></td>
    </tr>
    <tr>
      <th>sreformatpath</th>
      <td><p>File path for installed sreformat (default: sreformat)</p></td>
    </tr>
    <tr>
      <th>patmanpath</th>
      <td><p>File path for installed patman (default: patman)</p></td>
    </tr>
    <tr>
      <th>patmanoutput</th>
      <td><p>TXT file containing patman output (saved as output_prefix + patman_seed.txt). If supplied, analysis begins at patman output processing stage.</p></td>
    </tr>
    <tr>
      <th>minvalidationentries</th>
      <td><p>Minimum number of entries for a validation category to be considered separately in statistical analysis (default: 40)</p></td>
    </tr>
    <tr>
      <th>num_runs</th>
      <td><p>Number of subsamples to create (default: 100)</p></td>
    </tr>
    <tr>
      <th>check_python</th>
      <td><p>Whether the Python version should be checked (default: T)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>CSV containing full positive and negative sets. Folder statistical_analysis of heatmaps showing significance of various features under Fisher exact and Chi-squared tests. Seed analysis will always be run, full miRNA and flanking analysis if the respective parameters are set to 1. Folder subsamples containing CSVs for 100 subsamples with positive and negative samples equal for use in classifiers and feature selection.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function saves various files (using specified output_prefix) and if you wish to start preparation using one of these pre-output files then these can be specified and preparation will skip to that point (this should only be done with files output by the function).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>preparedataset</span><span class='op'>(</span>
   pythonversion<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/Sys.which.html'>Sys.which</a></span><span class='op'>(</span><span class='st'>'python'</span><span class='op'>)</span>,
   positiveset <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>'samples'</span>,<span class='st'>'test_seed_positive.csv'</span>,package<span class='op'>=</span><span class='st'>'feamiR'</span><span class='op'>)</span>,
   negativeset<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>'samples'</span>,<span class='st'>'test_seed_negative.csv'</span>,package<span class='op'>=</span><span class='st'>'feamiR'</span><span class='op'>)</span>,
   o<span class='op'>=</span><span class='st'>'examples_'</span>,
   num_runs<span class='op'>=</span><span class='fl'>0</span>,
   check_python<span class='op'>=</span><span class='st'>'F'</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: '-W' not found</span></div><div class='output co'>#&gt; [1] 127</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Eleanor Williams, Irina Mohorianu.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


